chia/plotters/plotters_util.py

```diff
@@ -0,0 +1,48 @@
+import asyncio
+import subprocess
+
+
+# https://kevinmccarthy.org/2016/07/25/streaming-subprocess-stdin-and-stdout-with-asyncio-in-python/
+async def _read_stream(stream, callback):
+    while True:
+        line = await stream.readline()
+        if line:
+            callback(line)
+        else:
+            break
+
+
+def parse_stdout(out, progress):
+    out = out.rstrip()
+    print(out, flush=True)
+    for k, v in progress.items():
+        if k in out:
+            print(f"Progress update: {v}", flush=True)
+
+
+async def run_plotter(args, progress_dict):
+    process = await asyncio.create_subprocess_exec(
+        *args, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
+    )
+
+    await asyncio.wait(
+        [
+            _read_stream(
+                process.stdout,
+                lambda x: parse_stdout(x.decode("UTF8"), progress_dict),
+            ),
+            _read_stream(
+                process.stderr,
+                lambda x: print("STDERR: {}".format(x.decode("UTF8"))),
+            ),
+        ]
+    )
+
+    await process.wait()
```

Comment:
#### --- 1

I'm not clear after a quick look over the docs as to what happens if a process fails and returns a non-zero exit code.  Does our code presently become aware of that?  Do we notify the user?  The docs didn't even make it clear if we could pass `check=True` to `asyncio.create_subprocess_exec()`...  `:|`  Otherwise I guess we just manually check the [`process.returncode`](https://docs.python.org/3.9/library/asyncio-subprocess.html#asyncio.asyncio.subprocess.Process.returncode).


#### --- 2

But, the docs did yield something I wasn't looking for.

https://docs.python.org/3.9/library/asyncio-subprocess.html#asyncio.asyncio.subprocess.Process.wait
> Note This method can deadlock when using `stdout=PIPE` or `stderr=PIPE` and the child process generates so much output that it blocks waiting for the OS pipe buffer to accept more data. Use the [`communicate()`](https://docs.python.org/3.9/library/asyncio-subprocess.html#asyncio.asyncio.subprocess.Process.communicate) method when using pipes to avoid this condition.

I didn't check in detail but I think the stream reading code that runs before this would do so until the streams are closed thus leaving no remaining opportunity for this deadlock to occur.  But, perhaps it would be just as well to use `.communicate()` instead.

Reply:
Regarding your first point, I tested a failed command that uses run_command, and indeed the exception is caught and our `exc_description` is printed, along with the command invocation and exit code.

I haven't observed any problems with the stream handling code while plotting with either plotter.